### ISSUE TITLE
Syntactic and Semantic Equivalencies

### ISSUE THREAD

**Issue by lyndonhenry** [souffle-lang/souffle#260]
Currently, Souffle does not recognize some of even the most trivial cases of equivalence. 

As pointed out in issue #42, identical rules/relations are not eliminated, for example

~~~
A(x,y) :- B(x,z), C(z,y).
A(x,y) :- B(x,z), C(z,y).
~~~

creates two distinct rules in the final program.

However, there are many similar cases that Souffle does not recognize, that are relatively simple. 

For example:

~~~
A(x,y) :- B(x,z), C(z,y).
A(a,b) :- B(a,c), C(c,b).
~~~

~~~
A(x,y) :- B(x,z), C(z,y).
A(x,y) :- C(z,y), B(x,z).
~~~

etc.

Ideally, it would be good to reduce these relations as much as possible at compile time.

**Comment by b-scholz** [comment]
This is related to the issue #42.

**Comment by azreika** [comment]
Implemented by PR #697.
